python/ray/tune/impl/tuner_internal.py

```diff
@@ -163,8 +164,58 @@ def fit(self) -> ResultGrid:
 
         return ResultGrid(analysis)
 
-    def _get_tune_run_arguments(self) -> Dict[str, Any]:
+    def _get_tune_run_arguments(self, trainable) -> Dict[str, Any]:
         """Get tune.run arguments common for both new and resumed runs."""
+        checkpoint_freq = self._run_config.checkpoint_config.checkpoint_frequency
+        checkpoint_at_end = self._run_config.checkpoint_config.checkpoint_at_end
+
+        if checkpoint_freq:
+            # Function trainables (and thus most of our trainers) usually don't handle
+            # this argument.
+            handle_checkpoint_freq = getattr(
+                trainable, "_handles_checkpoint_freq", None
+            )
+            if handle_checkpoint_freq is False:
+                # If we specifically know this trainable doesn't support the
+                # argument, raise an error
+                raise ValueError(
+                    f"You passed `checkpoint_freq={checkpoint_freq}` to your "
+                    f"CheckpointConfig, but this trainer does not support "
+                    f"this argument. If the trainer takes in a training loop, "
+                    f"you will need to trigger checkpointing yourself using "
+                    f"`ray.air.session.report(metrics=..., checkpoint=...)`."
+                )
+            elif handle_checkpoint_freq is True:
+                # If we specifically support it, it's handled in the training loop,
+                # so we disable tune's bookkeeping.
+                checkpoint_freq = 0
+            # Otherwise, this is a non-trainer trainable and we just keep the
+            # user-supplied value.
+
+        if checkpoint_at_end is not None:
+            # Again, function trainables usually don't handle this argument.
+            handle_cp_at_end = getattr(trainable, "_handles_checkpoint_at_end", None)
+            if handle_cp_at_end is False:
+                # If we specifically know we don't support it, raise an error.
+                raise ValueError(
+                    f"You passed `checkpoint_at_end={checkpoint_at_end}` to your "
+                    f"CheckpointConfig, but this trainer does not support "
+                    f"this argument. If the trainer takes in a training loop, "
+                    f"you will need to trigger checkpointing yourself using "
+                    f"`ray.air.session.report(metrics=..., checkpoint=...)`. "
+                )
+            elif handle_cp_at_end is True:
+                # If we specifically support it, it's handled in the training loop,
+                # so we disable tune's internal bookkeeping.
+                checkpoint_at_end = False
+            # If this is a user-defined trainable, just keep the value
+        elif checkpoint_at_end is None:
+            # Set default to False for function trainables and True for everything else
```

Comment:
intuitively, why does the type of trainable has anything to do with whether we should checkpoint at end??
are these 2 orthogonal things?

Reply:
Function trainables and generic training loop trainers like the TorchTrainer don't support driver-based checkpointing at the end of a run. This is because the user defines the training loop and thus decides themselves where to store checkpoints. This is in contrast to e.g. rllib where we can just call `trainable.save.remote()` anytime we want

Reply:
right, I guess I am still not sure why we want to default checkpoint_at_end to True just because we are able to do it for class trainables.